doc/automated-testing.md

```diff
@@ -0,0 +1,35 @@
+# Automated Testing
+
+## API Tests
+
+### Fake Modules
+
+The API tests use a fake implementation of interfaces that are used to fetch 
+or update records
+
+These fakes are configured in the FakeOkapi class.
+
+#### Schema Validation
+
+In order to try to identify potentially mismatches between the fake modules and 
+the reference implementations, it is possible to configure them to use a copy of
+an interface JSON schema to validate incoming representations during creation or 
+update of records.
+
+##### Location of Schemas
+
+Copies of the relevant storage schema are kept in the test resources directory.
+
+They are kept in the root in order to make it easier for schema references to be
+resolved, as the shared RAML definitions are included in the main resources
+
+They are named after the record and interface version, for example, the loan 
+record from the `loan-storage 6.1` interface is named `storage-loan-6-1`.
+
+##### Updating a Schema
+
+When an interface dependency is updated, any schema that are used within the 
+tests should also be updated, other the tests might produce false results.
```

Comment:
Another question is when do we do it? When the dependent module is released or when the change occurs, knowing more changes could occur, including complete rollback, before the release?

This is something that Pact (consumer contract validation) does by using a registry. I'm not sure we need to go this route, but, absent of a central authority, we'll need to be careful to ensure the copies of interfaces are correct.

Reply:
These are good questions, and representative of the kinds of questions that deterred me from introducing this kind complication to changes to integration points.

I agree that the documentation is likely insufficient.

> How do you recommend us knowing/remembering to update this dependency in mod-circulation?

I'm not sure I follow. Updating the interface dependency should happen either when mod-circulation needs something in a new minor version or it has to, in order to support a new major version.

If by dependency, you mean the copy of the schema from the storage interface, then I don't have a good answer. And that is what makes introducing this complicated.

As these changes should be driven by the need to change the interface dependency, that is what should drive it. Applying the production changes without updating the copy of the storage schema should cause failures, if it doesn't then it suggests the dependency change doesn't have any impact.

> Another question is when do we do it? When the dependent module is released or when the change occurs, knowing more changes could occur, including complete rollback, before the release?

I don't think I follow your reasoning. To me, this should be done when the interface dependency changes. 

If the dependency change is rolled back, then the production code changes, copy of the storage schema, tests etc also need to be rolled back too.

How does formal releases impact this decision?

> This is something that Pact (consumer contract validation) does by using a registry. I'm not sure we need to go this route, but, absent of a central authority

How does a central authority help with this, because it could be possible to ask it for the definition of  the interface e.g. `loan-storage 4.4`?

> we'll need to be careful to ensure the copies of interfaces are correct.

I agree that we need to be careful that the copies are correct, and that is a risk of this approach.

That is why I decided that the interface version number should be in the name of the schema, in order to try to make discrepancies more obvious. Though this is rather a weak signal.

Reply:
> I don't think I follow your reasoning. To me, this should be done when the interface dependency changes.

That means we will have to monitor every project mod-circulation uses for schema changes. That seems like a tall task. Even minor version changes could cause breaking changes in the client come integration time. I guess this is doable, but it seems like we'd need to have better monitoring of the schemas/interfaces changes for all the dependencies so we don't miss anything.

> How does a central authority help with this, because it could be possible to ask it for the definition of the interface e.g. loan-storage 4.4?

My concern is maintainability. I'm not a big fan of the schemas being tied to the project, but it is what it is. Having multiple copies seems like discrepancies could arise. Maybe I am too oversensitive to this.

Reply:
> > I don't think I follow your reasoning. To me, this should be done when the interface dependency changes.
> 
> That means we will have to monitor every project mod-circulation uses for schema changes. That seems like a tall task. Even minor version changes could cause breaking changes in the client come integration time. I guess this is doable, but it seems like we'd need to have better monitoring of the schemas/interfaces changes for all the dependencies so we don't miss anything.
> 

Why do we need to monitor every interface that mod-circulation uses?

What I don't really understand is how this changes the current challenges around synchronising interface changes. Can you help me understand why we need to do that more broadly?

From my perspective, we already need to synchronise major breaking compatibility changes of interface dependencies, in order to provide implementation versions that can work together (and keep the reference environments updating).

We also already need to update the minor interface versions when we need the new properties introduced. When we don't need those properties, we don't update the interface dependency, so we don't need to change the copied schema either. 

To me, this validation is checking that the representations produced by mod-circulation are compatible with the version of the interface it states support for. If a future minor change actually breaks compatibility, that is a broader challenge, and we likely need different mechanisms to detect that.

The only extra step, I think this change introduces is the need to change the schema copy (and only for the schema we especially care about in the tests). What am I missing?

> > How does a central authority help with this, because it could be possible to ask it for the definition of the interface e.g. loan-storage 4.4?
> 
> My concern is maintainability. I'm not a big fan of the schemas being tied to the project, but it is what it is. Having multiple copies seems like discrepancies could arise. Maybe I am too oversensitive to this.

I am also concerned about maintainability with this approach. Discrepancies are a risk with this approach (which is what drove my caution at doing this). 

I don't really have an alternative at this point. Do you?

Keeping the interface definitions with the reference implementations was an expedient decision a long time ago, in part because we couldn't come agree on a good centralised model. It involves coming up change management and governance structures, potentially restricting changes, deciding which versions to keep etc.

FOLIO likely does need to revisit this topic at some point, especially when it starts to bring in broader community contribution.

I think there are some significant challenges around version management (e.g. do we keep every version of the schema ever, in order to avoid congruence problems) and governance.

Reply:
> Why do we need to monitor every interface that mod-circulation uses?

How else do we know changes were made to the interfaces mod-circulation depends on? Especially changes that aren't announced via an interface version change (not following process) or changes minor changes that should be compatible that are not actually.

> What I don't really understand is how this changes the current challenges around synchronising interface changes. Can you help me understand why we need to do that more broadly?

There is no change in the challenges, I am just wondering how strict we need to be. That is, how heavily do we rely on this to catch breaking changes, expected or otherwise, at test time.

Reply:
> > Why do we need to monitor every interface that mod-circulation uses?
> 
> How else do we know changes were made to the interfaces mod-circulation depends on? Especially changes that aren't announced via an interface version change (not following process) or minor changes that should be compatible that are not actually.

I think these are separate challenges.

This change is to check that the output of mod-circulation complies with what it expects to produce (and what it explicitly declares a dependency on).

The challenge I think you are describing is how to do interface management across FOLIO. In terms of compliance to semantic versioning and coordinating changes from providers and consumers.

Am I understanding this correctly?

If so, I agree that is an important topic and it needs addressing more broadly in a different way.

> 
> > What I don't really understand is how this changes the current challenges around synchronising interface changes. Can you help me understand why we need to do that more broadly?
> 
> There is no change in the challenges, I am just wondering how strict we need to be. That is, how heavily do we rely on this to catch breaking changes, expected or otherwise, at test time.

I don't follow how this approach would catch the kinds of issues you refer to?

To me, this is about checking that mod-circulation is meeting it's obligations, rather than trying to spot when the outside environment has changed. I don't know how any isolated tests could do this? 

(aside: being able to check real dependencies is one of the nice side effects of being able to run these tests against real implementations, however that was lost at the beginning of the year)

Reply:
OK. I guess I misunderstood the intent here. I don't seem to be following what is going on and am on vacation, so I don't have much time to put into understanding the use case. I don't want to hold this up, so I'll defer to Martin and others for approval and will try to understand this code next week when I am back in the office.